File: core/core-mvc/src/main/java/io/glory/coremvc/filter/ContentCachingFilter.java
package io.glory.coremvc.filter;

import java.io.IOException;

import jakarta.annotation.PostConstruct;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.annotation.WebFilter;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import lombok.extern.slf4j.Slf4j;

import io.glory.coremvc.filter.servletwrapperprovider.CachedBodyHttpServletRequest;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.util.Assert;
import org.springframework.web.filter.OncePerRequestFilter;
import org.springframework.web.util.ContentCachingResponseWrapper;
import org.springframework.web.util.WebUtils;

/**
 * Request/Response Body 를 Caching 하는 Filter
 */
@WebFilter(filterName = "ContentCachingFilter", urlPatterns = "/**")
@Order(value = Ordered.HIGHEST_PRECEDENCE)
@Slf4j
public class ContentCachingFilter extends OncePerRequestFilter {

    @PostConstruct
    private void init() {
        log.info("# ==> ContentCachingFilter initialized");
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {

        var cachedBodyHttpServletRequest = new CachedBodyHttpServletRequest(request);
        var contentCachingResponse = new ContentCachingResponseWrapper(response);

        // ==============================
        // doFilter
        // ==============================
        filterChain.doFilter(cachedBodyHttpServletRequest, contentCachingResponse);

        ContentCachingResponseWrapper wrapper =
                WebUtils.getNativeResponse(contentCachingResponse, ContentCachingResponseWrapper.class);

        Assert.notNull(wrapper, "wrapper response is null");
        wrapper.copyBodyToResponse();
    }

}

File: core/core-mvc/src/main/kotlin/io/glory/coremvc/filter/ContentCachingFilter.kt
package io.glory.coremvc.filter

import io.github.oshai.kotlinlogging.KotlinLogging
import io.glory.coremvc.FILTER_EXCLUDE_PATH
import io.glory.coremvc.filter.servletwrapperprovider.CachedBodyHttpServletRequest
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.annotation.WebFilter
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.util.Assert
import org.springframework.web.filter.OncePerRequestFilter
import org.springframework.web.util.ContentCachingResponseWrapper
import org.springframework.web.util.WebUtils
import java.io.IOException

private val log = KotlinLogging.logger {}

/**
 * Content caching filter
 * <p>This filter is used to cache the request body and response body.</p>
 */
@WebFilter(filterName = "ContentCachingFilter", urlPatterns = ["/**"])
class ContentCachingFilter : OncePerRequestFilter() {

    init {
        log.info { "# ==> ContentCachingFilter initialized" }
    }

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val cachedBodyHttpServletRequest = CachedBodyHttpServletRequest(request)
        val contentCachingResponse = ContentCachingResponseWrapper(response)

        filterChain.doFilter(cachedBodyHttpServletRequest, contentCachingResponse)

        val wrapper = WebUtils.getNativeResponse(contentCachingResponse, ContentCachingResponseWrapper::class.java)

        Assert.notNull(wrapper, "wrapper response is null")
        wrapper!!.copyBodyToResponse()
    }

    @Throws(ServletException::class)
    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        val path = request.requestURI
        return FILTER_EXCLUDE_PATH.any { exclusion -> path.startsWith(exclusion) || path.endsWith(exclusion) }
    }

}
